new(modern_bpf): add readv,preadv,pread syscall

loresuso opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

/area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:
Support for `readv`, `preadv`,`pread` syscall in modern bpf

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Part of #723 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
NONE
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/917#" title="Author self-approved">loresuso</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign gnosek for approval. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["gnosek"]} -->

loresuso commented:
Ready for review!

loresuso commented on driver/modern_bpf/helpers/store/auxmap_store_params.h:
```diff
@@ -1137,6 +1128,30 @@ static __always_inline void auxmap__store_iovec_data_param(struct auxiliary_map
+ * @param msghdr_pointer pointer to `user_msghdr` struct.
+ * @param len_to_read imposed snaplen.
+ */
+static __always_inline void auxmap__store_msghdr_iovec_data_param(struct auxiliary_map *auxmap, unsigned long msghdr_pointer, unsigned long len_to_read)
+{
+	/* Read the usr_msghdr struct into the stack, if we fail,
+	 * we return an empty param.
+	 */
+	struct user_msghdr msghdr = {0};
+	if(bpf_probe_read_user((void *)&msghdr, bpf_core_type_size(struct user_msghdr), (void *)msghdr_pointer))
+	{
+		push__param_len(auxmap->data, &auxmap->lengths_pos, 0);
+		return;
+	}
+
+	u32 iov_cnt = msghdr.msg_iovlen;
+
+	auxmap__store_iovec_data_param(auxmap, (unsigned long)msghdr.msg_iov, iov_cnt, len_to_read);
+}
```
Note for the reviewer: the message header structure `msghdr` wraps a pointer to an array of `iovec`. Let's just isolate the handling of `iovec`, so that it can be reused directly in `readv` fillers.

loresuso commented:
I'm adding also `preadv` and `pread` soon, since they are similar

loresuso commented:
Added all three syscalls!

incertum commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/preadv.bpf.c:
```diff
@@ -0,0 +1,104 @@
+		if(bytes_to_read > ret)
+		{
+			bytes_to_read = ret;
+		}
```
@loresuso @hbrueckner and @Andreagit97 have a question I was holding back when we did the refactor wrt to introducing some upper bounds when applicable. Are there any risks associated with just trusting `ret`?

@loresuso please note that this comment is not related to this PR and rather a general question.

hbrueckner replied:
> @loresuso @hbrueckner and @Andreagit97 have a question I was holding back when we did the refactor wrt to introducing some upper bounds when applicable. Are there any risks associated with just trusting `ret`?

`ret` typically comes from the kernel implementation of the system call. Not quite sure if you can modify it (before exit) with e.g. `kprobes` or `ptrace

loresuso replied:
This is an interesting question! Do not know of any ways to do that in general, but I do agree with @hbrueckner that this is something that comes from the kernel and we can trust. About modification, I guess it is possible somehow, but doing so would mean already having high privileges on the system for sure, so I wouldn't worry too much

incertum replied:
Thanks @loresuso and @hbrueckner! Lastly we also never expect the size to be super large right and some safeguard upper bound is not needed here?

loresuso replied:
That check should already be in place, otherwise the verifier could go crazy :) We are going to use `push__bytebuf` which in turn uses `bpf_probe_read_{kernel,user}` and `SAFE_ACCESS` macro to limit which implements the upper bound to `MAX_PARAM_SIZE`

Andreagit97 replied:
agree with @loresuso and @hbrueckner here, `ret` comes directly from the kernel if someone alters it, it means that the kernel is already compromised so probably not an issue

incertum reviewed (approved):
Thanks @loresuso also for the nice cleanups throughout. Primarily checked for consistency and everything looks fine. Another reviewer should give it another close look please. 

Only have a question unrelated to this PR.

/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 39cfbc1a308b368607859a9424eb46d8f6e14ea1</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1322855705" title="Approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

hbrueckner commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/recvmsg.bpf.c:
```diff
@@ -80,7 +80,7 @@ int BPF_PROG(recvmsg_x,
 		/* Parameter 3: data (type: PT_BYTEBUF) */
 		unsigned long msghdr_pointer = extract__syscall_argument(regs, 1);
-		auxmap__store_iovec_data_param(auxmap, msghdr_pointer, bytes_to_read);
+		auxmap__store_msghdr_iovec_data_param(auxmap, msghdr_pointer, bytes_to_read);
```
self-review: this change will then be also required in `socketcall` PR #811 (depending which one gets merged earlier).

hbrueckner commented on test/drivers/test_suites/syscall_exit_suite/pread_x.cpp:
```diff
@@ -0,0 +1,50 @@
+
+	int32_t fd = -1;
+    size_t size = 13;
+    off_t pos = 17;
```
minor: could you check leading whitespace vs. tabs.

loresuso replied:
wops thanks!

poiana commented:
New changes are detected. LGTM label has been removed.

FedeDP commented:
/milestone next-driver

incertum reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 972de1957b7e7ebdd5ff4dd504456870a433ec57</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1326622174" title="Approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 commented on driver/fillers_table.c:
```diff
@@ -111,24 +111,20 @@ const struct ppm_event_entry g_ppm_events[PPM_EVENT_MAX] = {
 	[PPME_SYSCALL_WRITEV_E] = {FILLER_REF(sys_writev_e)},
 	[PPME_SYSCALL_WRITEV_X] = {FILLER_REF(sys_writev_pwritev_x)},
 #ifdef CAPTURE_64BIT_ARGS_SINGLE_REGISTER
-	[PPME_SYSCALL_PREADV_E] = {FILLER_REF(sys_autofill), 2, APT_REG, {{0}, {3} } },
+	[PPME_SYSCALL_PREADV_E] = {FILLER_REF(sys_preadv_e)},
 #else
 	[PPME_SYSCALL_PREADV_E] = {FILLER_REF(sys_preadv64_e)},
 #endif
```
could we use the same approach of `pread` removing the ifdef and using the same filler?

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -722,14 +757,28 @@ FILLER(sys_readv_preadv_x, true)
+	/*
+	* data and size
+	*/
+	if (retval > 0)
```
Here we need to choose if we want to discard the payload in case of failure, same question we have here  https://github.com/falcosecurity/libs/pull/900/files#r1113393171

Andreagit97 replied:
probably the best approach would be to get all the data also when the syscall fails, we will read junk in most of the cases but we could also catch some useful data in case of strange attack patterns

Andreagit97 replied:
btw in this PR you can keep it as it, we will adapt all the case in a separate one since we have to adapt all possible cases

loresuso replied:
Agreed in doing this in a separate PR

Andreagit97 commented on driver/ppm_fillers.c:
```diff
@@ -3867,6 +3875,48 @@ int f_sys_preadv64_e(struct event_filler_arguments *args)
+	return add_sentinel(args);
+}
+
+int f_sys_preadv_e(struct event_filler_arguments *args)
```
Same as BPF could we reuse the same filler `f_sys_preadv64_e` as you have done with `f_sys_pread64_e`?

Andreagit97 commented:
/hold

poiana commented:
New changes are detected. LGTM label has been removed.

loresuso commented:
I have rebased this PR and solved conflicts with the `socketcall` one (pay attention), and addressed @Andreagit97 comment, nice catch! Btw, I am not getting why `ci/circleci: test-drivers-engines-arm64` is failing 🤔

Andreagit97 commented on driver/fillers_table.c:
```diff
@@ -111,27 +111,19 @@ const struct ppm_event_entry g_ppm_events[PPM_EVENT_MAX] = {
-#else
-	[PPME_SYSCALL_PREADV_E] = {FILLER_REF(sys_preadv64_e)},
-#endif
+	[PPME_SYSCALL_PREADV_E] = {FILLER_REF(sys_preadv_e)},
```
so `sys_preadv_e` now replace `sys_preadv64_e` if I understood correctly... so could we remove `sys_preadv64_e`?

Andreagit97 replied:
probably I would rename also `sys_pread64_e` into `sys_pread_e` if it is ok for you

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -738,6 +738,41 @@ static __always_inline int bpf_parse_readv_writev_bufs(struct filler_data *data,
+	return bpf_val_to_ring(data, (int64_t)fd);
+}
+
+FILLER(sys_preadv_e, true)
```
if we use just that filler for both 64 bit and 32 bit capture we need the 
```c
#ifndef CAPTURE_64BIT_ARGS_SINGLE_REGISTER
#error Implement this
#error Implement this
#endif
#endif
```

like in `sys_pread64_e`

Andreagit97 reviewed (approved):
Thank you for this!
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 8266ebaa0a540aae33cd0e7db5d0821f6a4f3f3a</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1336807754" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1326622174" title="Approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1336807754" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1336830760" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1326622174" title="Approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

loresuso commented:
/unhold

loresuso commented:
/close 
/reopen

poiana commented:
@loresuso: Closed this PR.

<details>

In response to [this](https://github.com/falcosecurity/libs/pull/917#issuecomment-1466041989):

>/close 
>/reopen


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

loresuso commented:
/reopen

poiana commented:
@loresuso: Reopened this PR.

<details>

In response to [this](https://github.com/falcosecurity/libs/pull/917#issuecomment-1466042530):

>/reopen
>


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1336807754" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1336830760" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1337020871" title="Approved">hbrueckner</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#pullrequestreview-1326622174" title="Approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/917#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
